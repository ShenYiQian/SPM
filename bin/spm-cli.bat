set DEBUG=SPM:* & node ./build/bin/spm.js publish -i ./test/book