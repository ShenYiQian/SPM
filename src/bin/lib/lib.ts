import * as LibPath from "path";
import * as LibFs from "mz/fs";
import * as LibMkdirP from "mkdirp";
import * as bluebird from "bluebird";
import * as http from "http";
import * as recursive from "recursive-readdir";
import {ResponseSchema} from "../../lib/ApiBase";

export enum RequestMethod { post, get }

export interface SpmConfig {
    host: string;
    port: number;
    secret: string;
}

export interface SpmPackageConfig {
    name: string;
    version: string;
    description?: string;
    dependencies?: {
        [key: string]: string;
    };
}

export interface SpmPackage extends SpmPackageConfig {
    dependenciesChanged?: {
        [key: string]: string;
    };
    downloadUrl?: string;
    isDependencies?: boolean;
}

export interface SpmPackageMap {
    [dirname: string]: SpmPackage;
}

export const mkdir = bluebird.promisify<string>(LibMkdirP) as {
    <T>(path: string): bluebird<T>
};

export const rmdir = (dirPath: string) => {
    let files = LibFs.readdirSync(dirPath);
    if (files.length > 0) {
        for (let i = 0; i < files.length; i++) {
            let filePath = LibPath.join(dirPath, files[i]);
            if (LibFs.statSync(filePath).isFile()) {
                LibFs.unlinkSync(filePath);
            } else {
                rmdir(filePath);
            }
        }
    }
    LibFs.rmdirSync(dirPath);
};

export namespace Spm {

    export const INSTALL_DIR_NAME: string = 'spm_protos';
    export const SPM_VERSION_CONNECTOR: string = '__v';
    export const SPM_ROOT_PATH: string = LibPath.join(__dirname, '..', '..', '..');

    /**
     * Find project dir
     *
     * @returns {string}
     */
    export function getProjectDir(): string {
        return process.cwd();
    }

    /**
     * Save secret value into .spmlrc
     *
     * @returns {void}
     */
    export function saveSecret(value: string) {
        let lrcPath = LibPath.join(Spm.getProjectDir(), '.spmlrc');
        LibFs.writeFileSync(lrcPath, value, 'utf-8');
    }

    /**
     * Load secret value from .spmlrc
     *
     * @returns {string}
     */
    export function loadSecret(): string {
        let lrcPath = LibPath.join(Spm.getProjectDir(), '.spmlrc');
        if (LibFs.existsSync(lrcPath) && LibFs.statSync(lrcPath).isFile()) {
            return LibFs.readFileSync(lrcPath).toString();
        } else {
            return '';
        }
    }

    /**
     * Get Spm config
     *
     * @returns {SpmConfig}
     */
    export function getConfig(path?: string): SpmConfig {
        let configPath = (path) ? path : LibPath.join(SPM_ROOT_PATH, 'config', 'config.json');
        if (LibFs.existsSync(configPath) && LibFs.statSync(configPath).isFile()) {
            return JSON.parse(LibFs.readFileSync(configPath).toString());
        } else {
            throw new Error('[Config] config file path have to be an absolute path!');
        }
    }

    /**
     * Read spm.json via config path
     *
     * @param {string} path
     * @returns {SpmPackageConfig}
     */
    export function getSpmPackageConfig(path: string): SpmPackageConfig {
        return JSON.parse(LibFs.readFileSync(path).toString());
    }

    /**
     * Get all installed SpmPackage from projectDir
     *
     * @returns {Promise<SpmPackageMap>}
     */
    export async function getInstalledSpmPackageMap(path?: string): Promise<SpmPackageMap> {
        let projectDir = path ? path : Spm.getProjectDir();
        let installDir = LibPath.join(projectDir, Spm.INSTALL_DIR_NAME);

        if (LibFs.existsSync(installDir) && LibFs.statSync(installDir).isDirectory()) {
            let files = await recursive(installDir, ['.DS_Store']);

            let spmPackageMap = {} as SpmPackageMap;
            for (let file of files) {
                let basename = LibPath.basename(file);
                if (basename.match(/.+\.json/) !== null) {
                    let dirname = LibPath.dirname(file).replace(installDir, '').replace('\\', '').replace('/', '');
                    let packageConfig = Spm.getSpmPackageConfig(file);
                    spmPackageMap[dirname] = {
                        name: packageConfig.name,
                        version: packageConfig.version,
                        dependencies: packageConfig.dependencies
                    };
                }
            }
            return spmPackageMap;
        } else {
            return {};
        }
    }

    /**
     * Replace string in file
     *
     * @param {string} filePath
     * @param {Array<[RegExp, any]>} conditions
     * @returns {Promise<void>}
     */
    export async function replaceStringInFile(filePath: string, conditions: Array<[RegExp, any]>) {
        try {
            let buffer = await LibFs.readFile(filePath);
            let content = buffer.toString();
            for (let [reg, word] of conditions) {
                content = content.replace(reg, word);
            }
            LibFs.writeFileSync(filePath, Buffer.from(content));
        } catch (e) {
            throw e;
        }
    }
}

export namespace SpmPackageRequest {
    /**
     * Get Spm Publish request config
     *
     * @returns {SpmConfig}
     */
    export function getRequestOption(path: string, method: RequestMethod = RequestMethod.get): http.RequestOptions {
        let spmHttpConfig = Spm.getConfig();
        return {
            host: spmHttpConfig.host,
            port: spmHttpConfig.port,
            method: RequestMethod[method],
            path: path,
        };
    }

    /**
     * Parse request response
     *
     * @param {string} chunk
     * @returns {any}
     */
    export function parseResponse(chunk: Buffer | string): any {
        let response = JSON.parse(chunk.toString()) as ResponseSchema;
        if (response.code < 0) {
            throw new Error(response.msg.toString());
        }
        return response.msg;
    }
}