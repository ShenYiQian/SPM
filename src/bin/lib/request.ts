import * as LibFs from "mz/fs";
import * as http from "http";
import * as qs from "querystring";
import {RequestMethod, SpmPackageRequest} from "./lib";

export function post(uri: string, params: Object, callback: (chunk: Buffer | string, resolve: any, reject: any) => void, handleResponse?: Function) {
    return new Promise((resolve, reject) => {
        // create request
        let reqOptions = SpmPackageRequest.getRequestOption(uri, RequestMethod.post);

        let req = http.request(reqOptions, (res) => {
            (handleResponse)
                ? handleResponse(res, resolve)
                : res.on('data', (chunk) => callback(chunk, resolve, reject));
        }).on('error', (e) => {
            reject(e);
        });

        // send content
        let reqParamsStr = qs.stringify(params);
        req.setHeader('Content-Type', 'application/x-www-form-urlencoded');
        req.setHeader('Content-Length', Buffer.byteLength(reqParamsStr, 'utf8').toString());
        req.write(reqParamsStr);
    });
}

export function postForm(uri: string, params: Object, filePath: Array<string>, callback: (chunk: Buffer | string, resolve: any, reject: any) => void) {
    return new Promise((resolve, reject) => {
        // create request
        let reqOptions = SpmPackageRequest.getRequestOption(uri, RequestMethod.post);
        let req = http.request(reqOptions, (res) => {
            res.on('data', (chunk) => callback(chunk, resolve, reject));
        }).on('error', (e) => {
            reject(e);
        });

        // send content
        let boundaryKey = Math.random().toString(16);
        let enddata = '\r\n----' + boundaryKey + '--';

        // build form params head
        let content = '';
        for (let key in params) {
            content += '\r\n----' + boundaryKey + '\r\n'
                + 'Content-Disposition: form-data; name="' + key + '" \r\n\r\n'
                + encodeURIComponent(params[key]);
        }

        let contentBinary = new Buffer(content, 'utf-8');
        let contentLength = contentBinary.length;

        // build upload file head
        if (filePath.length > 0) {
            content += '\r\n----' + boundaryKey + '\r\n'
                + 'Content-Type: application/octet-stream\r\n'
                + 'Content-Disposition: form-data; name="fileUpload"; filename="' + `${Math.random().toString(16)}.zip` + '"\r\n'
                + "Content-Transfer-Encoding: binary\r\n\r\n";
            contentBinary = new Buffer(content, 'utf-8');
            contentLength = LibFs.statSync(filePath[0]).size + contentBinary.length;

            // send request stream
            let fileStream = LibFs.createReadStream(filePath[0]);
            fileStream.on('end', () => {
                req.end(enddata);
            });
            fileStream.pipe(req, {end: false});
        }

        // send request headers
        req.setHeader('Content-Type', 'multipart/form-data; boundary=--' + boundaryKey);
        req.setHeader('Content-Length', `${contentLength + Buffer.byteLength(enddata)}`);
        req.write(contentBinary);

        if (filePath.length === 0) {
            req.end(enddata);
        }
    });
}