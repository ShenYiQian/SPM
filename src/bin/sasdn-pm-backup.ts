import * as LibFs from "mz/fs";
import * as LibPath from "path";
import * as program from "commander";
import * as archiver from "archiver";
import {mkdir, Spm} from "./lib/lib";

const pkg = require('../../package.json');
const debug = require('debug')('SPM:CLI:install');

program.version(pkg.version)
    .parse(process.argv);

const OUTPUT_PATH = program.args[0] === undefined ? undefined : program.args[0];

export class BackupCLI {

    private _projectDir: string;
    private _tmpDir: string;
    private _tmpFileName: string;

    static instance() {
        return new BackupCLI();
    }

    public async run() {
        debug('BackupCLI start.');
        await this._validate();
        await this._prepare();
        await this._compress();
        await this._backup();
    }

    private async _validate() {
        debug('BackupCLI validate.');

        if (!OUTPUT_PATH) {
            throw new Error('output path is required, "sasdn-pm backup /tmp"');
        }

        let protoStat = await LibFs.stat(OUTPUT_PATH);
        if (!protoStat.isDirectory()) {
            throw new Error('output path is not a directory');
        }
    }

    private async _prepare() {
        debug('PublishCLI prepare.');

        const now = new Date();

        this._projectDir = Spm.getProjectDir();
        this._tmpDir = LibPath.join(Spm.SPM_ROOT_PATH, 'tmp');
        this._tmpFileName = `SpmPackage_${now.getFullYear()}_${now.getMonth() + 1}_${now.getDate()}_${Math.random().toString(16)}.zip`;
        await mkdir(this._tmpDir);
    }

    private async _compress() {
        debug('BackupCLI compress.');

        let tmpFilePath = LibPath.join(this._tmpDir, this._tmpFileName);

        // create a file to stream archive data to.
        await new Promise(async (resolve, reject) => {
            // create write stream
            let writeStream = LibFs.createWriteStream(tmpFilePath)
                .on('close', () => {
                    resolve();
                });
            // archive init
            let archive = archiver('zip', {zlib: {level: 9}})
                .on('error', (err) => reject(err)) as archiver.Archiver;

            archive.pipe(writeStream);

            // add store dir
            let storePath = LibPath.join(this._projectDir, 'store');
            if (LibFs.existsSync(storePath) && LibFs.statSync(storePath).isDirectory()) {
                archive.directory(storePath, '/store');
            }

            // add Spm.db file
            let dbPath = LibPath.join(this._projectDir, 'Spm.db');
            if (LibFs.existsSync(dbPath) && LibFs.statSync(dbPath).isFile()) {
                archive.append(LibFs.createReadStream(LibPath.join(this._projectDir, 'Spm.db')), {name: 'Spm.db'});
            }

            await archive.finalize();
        });

        debug('BackupCLI compress finish.');
    }

    private async _backup() {
        debug('BackupCLI backup.');
        let sourceFile = LibPath.join(this._tmpDir, this._tmpFileName);
        let destFile = LibPath.join(OUTPUT_PATH, this._tmpFileName);

        await LibFs.rename(sourceFile, destFile, (err) => {
            if (err) throw err;
            LibFs.stat(destFile, (err) => {
                if (err) throw err;
                debug('BackupCLI backup finish.');
            });
        });
    }
}

BackupCLI.instance().run().catch((err: Error) => {
    debug('err: %O', err.message);
});