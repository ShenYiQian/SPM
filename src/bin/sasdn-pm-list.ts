import * as program from 'commander';
import * as LibPath from 'path';
import {Spm} from './lib/lib';

const pkg = require('../../package.json');

program.version(pkg.version)
    .parse(process.argv);

export class ListCLI {
    private _projectDir: string;

    static instance() {
        return new ListCLI();
    }

    public async run() {
        console.log('ListCLI start.');

        await this._prepare();
        await this._displayPackageList();

        console.log('ListCLI complete.');
    }

    /**
     * 准备命令中需要使用的参数，或创建文件夹。
     *
     * @returns {Promise<void>}
     * @private
     */
    private async _prepare() {
        console.log('ListCLI prepare.');

        this._projectDir = Spm.getProjectDir();
    }

    /**
     * 显示所有已安装的 package
     *
     * @returns {Promise<void>}
     * @private
     */
    private async _displayPackageList() {
        console.log('ListCLI show.');

        let spmPackageMap = await Spm.getInstalledSpmPackageMap();

        console.log('--------------Installed SpmPackage---------------');
        for (let dirname in spmPackageMap) {
            let spmPackage = spmPackageMap[dirname];
            console.log(`+-- ${spmPackage.name}@${spmPackage.version}`);
            for (let dependName in spmPackage.dependencies) {
                console.log(`|  | -- ${dependName}@${spmPackage.dependencies[dependName]}`);
            }
        }
        console.log('--------------Installed SpmPackage---------------');
    }

}

ListCLI.instance().run().catch((err: Error) => {
    console.log('error:', err.message);
});