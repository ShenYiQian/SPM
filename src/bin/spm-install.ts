import * as program from "commander";
import * as LibPath from "path";
import * as LibFs from "mz/fs";
import {findProjectDir, RequestMethod, rmdir, SpmHttp, SpmPackageInfoSchema, SpmPackageConfig} from "./lib/lib";
import * as http from "http";
import * as recursive from "recursive-readdir";
import * as unzip from "unzip";
import * as qs from "querystring";
import * as _ from "underscore";
import {ResponseSchema} from "../lib/Router";

const pkg = require('../../package.json');
const debug = require('debug')('SPM:CLI:install');

program.version(pkg.version)
    .option('-n, --name <item>', 'package name')
    .option('-p, --projectDir <dir>', 'project dir')
    .parse(process.argv);

const NAME_VALUE = (program as any).name === undefined ? undefined : (program as any).name;
const PROJECT_DIR_VALUE = (program as any).projectDir === undefined ? undefined : (program as any).projectDir;

class InstallCLI {
    private _tmpDir: string;
    private _tmpFileName: string;
    private _projectDir: string;
    private _projectProtoDir: string;
    private _projectInstalled: { [dirName: string]: [string, string, string] };
    private _installList: { [dirName: string]: SpmPackageInfoSchema };

    static instance() {
        return new InstallCLI();
    }

    public async run() {
        debug('InstallCLI start.');
        await this._validate();
        await this._initLocaleInfo();
        await this._initRemoteInfo();
        await this._deploy();
    }

    private async _validate() {
        debug('InstallCLI validate.');

        if (!NAME_VALUE) {
            throw new Error('--name is required');
        }

        // 向上查找项目文件夹根目录
        if (!PROJECT_DIR_VALUE) {
            this._projectDir = findProjectDir(__dirname);
        } else {
            this._projectDir = PROJECT_DIR_VALUE;
        }
    }

    private async _initLocaleInfo() {
        debug('InstallCLI init LocaleInfo.');

        // 创建临时文件夹
        this._tmpDir = LibPath.join(__dirname, '..', '..', 'tmp');
        this._tmpFileName = Math.random().toString(16) + ".zip";
        await mkdir(this._tmpDir);

        // 创建项目根目录下的spm_protos文件夹
        this._projectProtoDir = LibPath.join(this._projectDir, "spm_protos");
        this._projectInstalled = {};
        await mkdir(this._projectProtoDir);

        let protoDirStat = await LibFs.stat(this._projectProtoDir);
        if (protoDirStat.isDirectory()) {
            let files = await recursive(this._projectProtoDir, [".DS_Store"]);
            for (let file of files) {
                let basename = LibPath.basename(file);
                if (basename.match(/.+\.json/) !== null) {
                    let name = LibPath.dirname(file).replace(this._projectProtoDir, '').replace('\\', '').replace('/', '');
                    let packageOption = JSON.parse(LibFs.readFileSync(file).toString()) as SpmPackageConfig;
                    let [major, minor, patch] = packageOption.version.split('.');
                    this._projectInstalled[name] = [major, minor, patch];
                }
            }
        }
    }

    private async _initRemoteInfo() {
        debug('InstallCLI init remote info.');

        this._installList = {};

        // 创建临时文件夹
        await new Promise(async (resolve, reject) => {
            let reqParamsStr = qs.stringify({
                name: NAME_VALUE,
            });

            // create request
            let reqOptions = await SpmHttp.getRequestOption('/v1/depend', RequestMethod.post);
            let req = http.request(reqOptions, (res) => {
                res.on("data", (chunk) => {
                    let response = JSON.parse(chunk.toString()) as ResponseSchema;
                    let depends = response.msg;
                    if (_.isObject(depends)) {
                        for (let pkgName in depends as SpmPackageInfoSchema) {
                            let [name, version] = pkgName.split('@');
                            this._mergeInstallPackage(name, version, depends[pkgName].path, depends[pkgName].dependencies);
                        }

                        this._changeInstallDependencies();
                        resolve();
                    } else {
                        reject(new Error(chunk.toString()));
                    }
                });
            }).on('error', (e) => reject(e));

            // send request headers
            req.setHeader('Content-Type', 'application/x-www-form-urlencoded');
            req.setHeader('Content-Length', Buffer.byteLength(reqParamsStr, 'utf8').toString());
            req.write(reqParamsStr);
        });

    }

    private _mergeInstallPackage(name: string, version: string, path: string, dependencies: { [key: string]: string }, deepLevel: number = 0, originalName?: string): void {
        let [nextMajor, nextMinor, nextPatch] = version.split('.');

        if (this._projectInstalled.hasOwnProperty(name)) {
            let [curMajor, curMinor, curPath] = this._projectInstalled[name];
            if (nextMajor == curMajor) {
                if (nextMinor < curMinor || nextMinor == curMinor && nextPatch < curPath) {
                    // 依赖版本低于当前版本，不处理，其他情况都要重新下载
                } else {
                    this._projectInstalled[name] = [nextMajor, nextMinor, nextPatch];
                    this._installList[name] = {
                        name: (originalName) ? originalName : name,
                        version: [nextMajor, nextMinor, nextPatch],
                        path: path,
                        dependencies: dependencies
                    };
                }
            } else {
                if (deepLevel == 0) {
                    this._mergeInstallPackage(name + nextMajor, version, path, dependencies, 1, name);
                }
            }
        } else {
            this._projectInstalled[name] = [nextMajor, nextMinor, nextPatch];
            this._installList[name] = {
                name: (originalName) ? originalName : name,
                version: [nextMajor, nextMinor, nextPatch],
                path: path,
                dependencies: dependencies
            };
        }
    }

    private _changeInstallDependencies() {
        for (let name in this._installList) {
            this._installList[name].dependenciesChangeMap = {};
            for (let dependName in this._installList[name].dependencies) {
                let [dependMajor] = this._installList[name].dependencies[dependName].split('.');
                if (this._projectInstalled.hasOwnProperty(dependName)) {
                    let [installMajor] = this._projectInstalled[dependName];
                    if (installMajor != dependMajor) {
                        this._installList[name].dependenciesChangeMap[dependName] = dependName + dependMajor
                    }
                }
            }
        }
    }

    private async _deploy() {
        debug('InstallCLI deploy.');
        for (let name in this._installList) {
            await this._install(name, this._installList[name]);
        }
    }

    private async _install(name: string, info: SpmPackageInfoSchema) {
        debug('InstallCLI install. name: ' + name);

        let tmpName = name + this._tmpFileName;
        let tmpZipPath = LibPath.join(this._tmpDir, tmpName);
        let tmpPkgPath = LibPath.join(this._tmpDir, name);

        // download file
        await new Promise(async (resolve, reject) => {
            debug('InstallCLI download.');
            let reqParamsStr = qs.stringify({
                path: info.path,
            });

            let fileStream = LibFs.createWriteStream(tmpZipPath);

            // create request
            let reqOptions = await SpmHttp.getRequestOption('/v1/install', RequestMethod.post);
            let req = http.request(reqOptions, (res) => {
                if (res.headers['content-type'] == 'application/octet-stream') {
                    res.pipe(fileStream);
                    res.on("end", () => {
                        debug('InstallCLI download finish');
                        resolve();
                    });
                } else {
                    res.on("data", (chunk) => reject(new Error(chunk.toString())));
                }
            }).on('error', (e) => reject(e));

            // send request headers
            req.setHeader('Content-Type', 'application/x-www-form-urlencoded');
            req.setHeader('Content-Length', Buffer.byteLength(reqParamsStr, 'utf8').toString());
            req.write(reqParamsStr);
        });

        // unzip file
        await new Promise(async (resolve, reject) => {
            debug('InstallCLI unzip.');
            let fileStat = await LibFs.stat(tmpZipPath);
            if (fileStat.isFile()) {
                LibFs.createReadStream(tmpZipPath).pipe(unzip.Extract({ path: tmpPkgPath }).on("close", async () => {
                    debug('InstallCLI unzip finish');
                    await LibFs.unlink(tmpZipPath);
                    resolve();
                }));
            } else {
                await LibFs.unlink(tmpZipPath);
                reject(new Error("Download file corruption."))
            }
        });

        // change package name
        await new Promise(async (resolve, reject) => {
            debug('InstallCLI replace.');
            if (_.isEmpty(info.dependenciesChangeMap) && info.name == name) {
                resolve();
            } else {
                let files = await recursive(tmpPkgPath, ['.DS_Store']);
                let count = 0;
                await files.map(async (file: string) => {
                    count++;
                    if (LibPath.basename(file).match(/.+\.proto/) !== null) {
                        if (info.name != name) {
                            await this._replaceStringInFile(file, [
                                [new RegExp(`package ${info.name};`,"g"), `package ${name};`]
                            ]);
                        }

                        if (!_.isEmpty(info.dependenciesChangeMap)) {
                            for (let oldDependName in info.dependenciesChangeMap) {
                                let newDependName = info.dependenciesChangeMap[oldDependName];
                                await this._replaceStringInFile(file, [
                                    [new RegExp(`import "${oldDependName}/`,"g"), `import "${newDependName}/`],
                                    [new RegExp(`\\((${oldDependName}.*?)\\)`,"g"), (word) => word.replace(oldDependName, newDependName)],
                                    [new RegExp(` (${oldDependName}.*?) `,"g"), (word) => word.replace(oldDependName, newDependName)]
                                ]);
                            }
                        }

                        if (count == files.length) {
                            resolve();
                        }
                    }
                });
            }
        });

        await new Promise(async (resolve) => {
            debug('InstallCLI spm_proto dir.');
            let installDir = LibPath.join(this._projectProtoDir, name);
            await rmdir(installDir);
            await LibFs.rename(tmpPkgPath, installDir);
            resolve();
        })
    }

    private async _replaceStringInFile(filePath: string, conditions: Array<[RegExp, any]>) {
        try {
            let fileStat = await LibFs.stat(filePath);
            if (fileStat.isFile()) {
                let content = LibFs.readFileSync(filePath).toString();
                for (let [reg, word] of conditions) {
                    content = content.toString().replace(reg, word);
                }
                await LibFs.writeFile(filePath, Buffer.from(content), (err) => {
                    if (err) {
                        throw err;
                    }
                });
            }
        } catch (e) {
            throw e;
        }
    }
}

InstallCLI.instance().run().catch((err: Error) => {
    debug('err: %O', err.message);
});