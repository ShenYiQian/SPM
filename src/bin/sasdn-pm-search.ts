import * as program from 'commander';
import * as _ from 'underscore';
import * as request from './lib/request';
import {SpmPackageRequest} from './lib/lib';
import {SpmPackage} from '../lib/entity/SpmPackage';
import {SpmPackageVersion} from '../lib/entity/SpmPackageVersion';

const pkg = require('../../package.json');

program.version(pkg.version)
    .option('-i, --info', 'show proto info')
    .parse(process.argv);

const INFO_VALUE = (program as any).info === undefined ? undefined : (program as any).info;
const KEYWORD_VALUE = program.args[0] === undefined ? undefined : program.args[0];

export class SearchCLI {

    static instance() {
        return new SearchCLI();
    }

    public async run() {
        console.log('SearchCLI start.');
        await this._validate();
        await this._search();
    }

    private async _validate() {
        console.log('SearchCLI validate.');

        if (!KEYWORD_VALUE) {
            throw new Error('keyword is required');
        }
    }

    private async _search() {
        console.log('SearchCLI search.');

        await new Promise(async (resolve, reject) => {
            let params = {
                keyword: KEYWORD_VALUE,
                info: !!(INFO_VALUE)
            };

            request.post('/v1/search', params, (chunk, reqResolve, reqReject) => {
                try {
                    reqResolve(SpmPackageRequest.parseResponse(chunk));
                } catch (e) {
                    reqReject(e);
                }
            }).then((response: Array<SpmPackage | [SpmPackage, SpmPackageVersion]>) => {
                console.log('--------------Search Response---------------');
                if (response.length > 0) {
                    for (let packageInfo of response) {
                        if (_.isArray(packageInfo)) {
                            let [spmPackage, spmPackageVersion] = packageInfo as [SpmPackage, SpmPackageVersion];
                            console.log(`${spmPackage.name}@${spmPackageVersion.major}.${spmPackageVersion.minor}.${spmPackageVersion.patch}`);
                        } else {
                            console.log(`${packageInfo.name} | ${(packageInfo.description) ? packageInfo.description : 'no description'}`);
                        }
                    }
                } else {
                    console.log('package not found!');
                }
                console.log('--------------Search Response---------------');
                resolve();
            }).catch((e) => {
                reject(e);
            });
        });

    }
}

SearchCLI.instance().run().catch((err: Error) => {
    console.log('error:', err.message);
});