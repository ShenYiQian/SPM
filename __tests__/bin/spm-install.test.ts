import * as LibFs from 'mz/fs';
import * as LibPath from 'path';
import {InstallCLI} from "../../src/bin/spm-install";
import {Spm, rmdir} from "../../src/bin/lib/lib";

jest.mock('../../src/bin/lib/request');

describe('SpmInstall', async () => {

    let installCLI = InstallCLI.instance();

    beforeAll(async () => {
        process.chdir('demo');
        //await rmdir(LibPath.join(process.cwd(), Spm.INSTALL_DIR_NAME));

    });

    afterAll(async() => {
        process.chdir('..');
    });

    // test
    test('_prepare: prepare data and validate data is right', async () => {
        await installCLI['_prepare']();


        expect(installCLI['_tmpDir']).toMatchSnapshot();
        expect(installCLI['_tmpFileName']).toMatch(/.zip/);
        expect(installCLI['_projectDir']).toMatchSnapshot();
        expect(installCLI['_spmPackageInstallDir']).toMatchSnapshot();
        expect(installCLI['_spmPackageInstalledMap']).toMatchSnapshot();

        // test
        let files1 = await LibFs.readdir(LibPath.join(installCLI['_tmpDir'], '..'));
        expect(files1[files1.indexOf('tmp')]).toEqual('tmp');

        let files2 = await LibFs.readdir(LibPath.join(installCLI['_spmPackageInstallDir'], '..'));
        expect(files2[files2.indexOf(Spm.INSTALL_DIR_NAME)]).toEqual(Spm.INSTALL_DIR_NAME);
    });

    // test
    test('_install: validate install is finish', async () => {
        // await installCLI['_install']();
    });
});