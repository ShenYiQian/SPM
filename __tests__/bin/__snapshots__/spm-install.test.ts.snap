// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpmInstall _prepare: prepare data and validate data is right 1`] = `"E:\\\\data1\\\\www\\\\demo\\\\SASDN-PM\\\\tmp"`;

exports[`SpmInstall _prepare: prepare data and validate data is right 2`] = `"E:\\\\data1\\\\www\\\\demo\\\\SASDN-PM\\\\demo"`;

exports[`SpmInstall _prepare: prepare data and validate data is right 3`] = `"E:\\\\data1\\\\www\\\\demo\\\\SASDN-PM\\\\demo\\\\spm_protos"`;

exports[`SpmInstall _prepare: prepare data and validate data is right 4`] = `Object {}`;
